Allow tainted scripts to extract text from some fields

sysrqb opened the pull request:
#### 614b6dccb3ba73d2974dccecb6992b88e29e68c1
<pre>
Allow tainted scripts to extract text from some fields
<a href="https://bugs.webkit.org/show_bug.cgi?id=301157">https://bugs.webkit.org/show_bug.cgi?id=301157</a>
<a href="https://rdar.apple.com/161669359">rdar://161669359</a>

Reviewed by Wenson Hsieh.

The form control protection prevents tainted scripts from accessing the value
of all text form fields. That protection causes some web compatibility issues
that we can fix by slightly relaxing the restriction which is what I&apos;m doing in
this patch. We can let tainted scripts extract the value if that element was
created by a tainted script and it wasn&apos;t modified by user input.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/ScriptTrackingPrivacyTests.mm

* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::m_wasCreatedByTaintedScript):
(WebCore::m_wasChangedSinceLastFormControlChangeEvent): Deleted.
* Source/WebCore/html/HTMLFormControlElement.h:
(WebCore::HTMLFormControlElement::wasCreatedByTaintedScript const):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::value const):
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::value const):
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::didEditInnerTextValue):
(WebCore::HTMLTextFormControlElement::wasChangeEverUserEdit const):
(WebCore::HTMLTextFormControlElement::shouldApplyScriptTrackingPrivacyProtection const):
* Source/WebCore/html/HTMLTextFormControlElement.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScriptTrackingPrivacyTests.mm:
(TestWebKitAPI::(ScriptTrackingPrivacyTests, DirectFormFieldAccess)):

Canonical link: <a href="https://commits.webkit.org/302031@main">https://commits.webkit.org/302031@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ec2e341388d77e9673b15781d32ab92d666f74c7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127737 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47385 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38555 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135009 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79294 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129609 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/48013 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55916 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97230 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65133 "") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130685 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38389 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114418 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77712 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/93447a7c-3748-432d-8253-9b194ed59c4f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/37182 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32518 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78379 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108264 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32973 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137493 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54396 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41934 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105751 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54907 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110775 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105403 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26898 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50924 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29356 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52022 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54333 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60514 "Built successfully") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53569 "Built successfully") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/57024 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/55326 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/html/HTMLTextFormControlElement.cpp:
```diff
@@ -155,6 +155,7 @@ void HTMLTextFormControlElement::didEditInnerTextValue(bool wasUserEdit)
     LOG(Editing, "HTMLTextFormControlElement %p didEditInnerTextValue", this);
 
     m_lastChangeWasUserEdit = wasUserEdit;
+    m_wasChangeEverUserEdit |= wasUserEdit;
```
```suggestion
    m_wasEverChangedByUserEdit |= wasUserEdit;
```

Maybe `m_wasEverChangedByUserEdit` or `m_wasEverUserEdited` might read a bit more nicely?

whsieh commented on Source/WebCore/html/HTMLTextFormControlElement.cpp:
```diff
@@ -947,4 +955,9 @@ void HTMLTextFormControlElement::adjustInnerTextStyle(const RenderStyle& parentS
+bool HTMLTextFormControlElement::shouldApplyScriptTrackingPrivacyProtection() const
+{
+    return protectedDocument()->requiresScriptTrackingPrivacyProtection(ScriptTrackingPrivacyCategory::FormControls)
+        && (wasChangeEverUserEdit() || !wasCreatedByTaintedScript());
```
I think if you make this `(wasChangeEverUserEdit() || !wasCreatedByTaintedScript()) && …`, we might be able to avoid the (potentially more expensive) script tracking privacy check.

whsieh commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -66,6 +66,7 @@ HTMLFormControlElement::HTMLFormControlElement(const QualifiedName& tagName, Doc
     , m_isRequired(false)
     , m_valueMatchesRenderer(false)
     , m_wasChangedSinceLastFormControlChangeEvent(false)
+    , m_wasCreatedByTaintedScript(document.requiresScriptTrackingPrivacyProtection(ScriptTrackingPrivacyCategory::FormControls))
```
Might be worth adding a new argument flag to `requiresScriptTrackingPrivacyProtection`, which would prevent us from adding a console log message in the case where it's tainted (or perhaps, a different message, that would make it clear that returning `true` here doesn't, by itself, block form access).

Commit:
Allow tainted scripts to extract text from some fields
https://bugs.webkit.org/show_bug.cgi?id=301157
rdar://161669359

Reviewed by Wenson Hsieh.

The form control protection prevents tainted scripts from accessing the value
of all text form fields. That protection causes some web compatibility issues
that we can fix by slightly relaxing the restriction which is what I'm doing in
this patch. We can let tainted scripts extract the value if that element was
created by a tainted script and it wasn't modified by user input.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/ScriptTrackingPrivacyTests.mm

* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::m_wasCreatedByTaintedScript):
(WebCore::m_wasChangedSinceLastFormControlChangeEvent): Deleted.
* Source/WebCore/html/HTMLFormControlElement.h:
(WebCore::HTMLFormControlElement::wasCreatedByTaintedScript const):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::value const):
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::value const):
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::didEditInnerTextValue):
(WebCore::HTMLTextFormControlElement::wasChangeEverUserEdit const):
(WebCore::HTMLTextFormControlElement::shouldApplyScriptTrackingPrivacyProtection const):
* Source/WebCore/html/HTMLTextFormControlElement.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScriptTrackingPrivacyTests.mm:
(TestWebKitAPI::(ScriptTrackingPrivacyTests, DirectFormFieldAccess)):

Canonical link: https://commits.webkit.org/302031@main
